essence-of-live-coding/test/Cell/Util.hs

```diff
@@ -168,4 +196,75 @@ test = testGroup "Utility unit tests"
     }
   ]
 
+type TestTraversables' = TestTraversables '[Maybe, [], V.Vector 10, Seq, Map Int]
+
+testTraverse' :: Test
+testTraverse' = testGroup "Traversing unit tests" 
+  [ genTraversableTests' @TestTraversables' "traverse' (arr f) = arr (f <$>)"
+    $ makeTraversableTest (traverseArrLaw @Int @Int)
+  , genTraversableTests' @TestTraversables'
+    "traverse' works as expected for any Cell Identy Int Int created with constructor Cell" 
+    $ makeTraversableTest (traverseCellTest @Int @Int @Int)
+  , testProperty "traverse' by itself does not force the entire list (ArrM)" $ CellSimulation
+    { cell = arr head
+    , input = [1 : error "Bang !"]
+    , output = [1] }
+  , testProperty "traverse' by itself does not force the entire list (Cell)" $ CellSimulation
+    { cell = cellId >>> arr head
+    , input = [1 : error "Bang !"]
+    , output = [1] }
+  ]  
+
+traverseArrLaw :: forall a b t . Traversable t 
+  => Proxy t
+  -> [t a]
+  -> Fun a b
+  -> CellIdentitySimulation (t a) (t b)
+traverseArrLaw _ input' (Fn f) = CellIdentitySimulation
+  { cell1' = arr (f <$>)
+  , cell2' = traverse' (arr f)
+  , .. }
+
+traverseCellTest :: forall s a b t . (Traversable t, Data s) 
+  => Proxy t -> s -> Fun (s, a) (b, s) -> [t a] ->
+  CellSimulation (t a) (t b)
+traverseCellTest _ s (Fn2 f) input = CellSimulation
+  { cell = traverse' (Cell s (\s a -> pure $ f s a))
+  , output = runIdentity $ evalStateT (traverse (traverse (\a -> StateT (Identity . (`f` a)) )) input) s
+  , ..
+  }
+
+makeTraversableTest :: forall (t :: * -> *) a . (Testable a, Typeable t) =>  (Proxy t -> a) -> [Char] -> Proxy t -> Test
+makeTraversableTest a message _ = testProperty (message <> " " <> show (typeRep (Proxy :: Proxy t))) (a (Proxy :: Proxy t))
+
+-- | A data type to store types which are instances of 'Traversable'.
+data TestTraversables :: [* -> *] -> *
+
+-- | A type class for induction on the type-level list containing the Traversables. 
+class GenTests a where
+  genTraversableTests ::  
+    (forall (t :: * -> *) . (Arbitrary (t Int), Show (t Int), Eq (t Int), Traversable t, Typeable t) => Proxy t -> Test)
```

Comment:
If I change all the occurences to Arbitrary1, Show1, Eq1, etc. And add

```Haskell
instance  (KnownNat n) => Arbitrary1 (V.Vector n) where
  liftArbitrary = V.replicateM
  liftShrink = V.mapM
instance (Arbitrary a, KnownNat n) => Arbitrary (V.Vector n a) where
  arbitrary = arbitrary1
  shrink = shrink1 
```

Then I get this error

```Haskell
test/Cell/Util.hs:231:9: error:
    • Could not deduce (Arbitrary (t Int))
        arising from a use of ‘makeTraversableTest’
      from the context: (Arbitrary1 t, Show1 t, Eq1 t, Traversable t,
                         Typeable t)
        bound by a type expected by the context:
                   forall (t :: * -> *).
                   (Arbitrary1 t, Show1 t, Eq1 t, Traversable t, Typeable t) =>
                   Proxy t -> Test
        at test/Cell/Util.hs:231:9-54
    • In the second argument of ‘($)’, namely
        ‘makeTraversableTest (traverseArrLaw @Int @Int)’
```
I don't entirely undestand how these type classes work (Show1, etc). Do you have an ideia of what need to do more ?

Reply:
I had another look at this, and I don't think that `Arbitrary1` will help much here, since there isn't an automatic instance for `Arbitrary (t Int)` even if there is an instance for `Arbitrary1 t`, and `Arbitrary Int`. Such an instance can be made for a specific t easilly (`arbitrary = arbitrary1`), but it is not made automatically it seems, as it would clash with other instances. So I think I would leave it as it is.